### ISSUE TITLE
passing non-fn literals to map methods

### ISSUE THREAD

**Issue by andrewbanchich** [rust-lang/rust-clippy#12214]
### What it does

Methods such as `Result::map_err` are supposed to be lazily evaluated, but Rust will not take any issue with code such as:

```rust
foo.map_err({
  println!("something bad happened");
  FooError::UhOh
})
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e44edfac03ae7f79fda7734555be9301)

This expression will be eagerly evaluated, and log that an error occurred even when no error occurred. I think it would be useful to have a lint to explain this to push people to only pass fn literals to methods which should be lazily evaluated.

### Advantage

Removes potential bugs the type system cannot catch.

### Drawbacks

I can't think of any good reason to pass anything other than a literal or a variable name.

### Example

```rust
foo.map_err({
  println!("something bad happened");
  FooError::UhOh
})
```

Could be written as:

```rust
foo.map_err(|e| {
  println!("something bad happened");
  FooError::UhOh(e)
})
```

**Comment by Centri3** [comment]
The suggestion could just be `|...|` before the block, and `(...)` at the end of the final expression (the path to the constructor) if it's difficult to properly suggest the arguments

I recall seeing a rustc issue for something like this, when it would fail to compile - This would also fit into rustc as clippy can't suggest adding `|_|` if it fails to compile (say, you pass a unit struct instead)

**Comment by asquared31415** [comment]
I'm not sure why this would ever be used in these specific cases of mapping, but note that a somewhat common pattern is
```rust
let my_arc = Arc::new(...);
use_closure({
    let my_arc = my_arc.clone();
    move || {
        /* use my_arc */
    }
});
```
to avoid creating extra variables in the parent scope when they are just going to be moved.

I don't think it's worth special casing this, but I would expect at least one user to raise an issue about this in the future.

**Comment by andrewbanchich** [comment]
Yeah, I was thinking about that too. It's also a common pattern for async code which accepts closures since it often needs to be move.

However, I'd think it could also be written as a closure defined outside the block and passed as a variable.

I'd also mention that if `use_closure` is meant to be lazy like `map_err`, some uses cases might consider cloning an Arc eagerly to be non-trivial, so this would also be helpful to be flagged.
